primitives/src/vstaging/executor_params.rs

```diff
@@ -0,0 +1,142 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Abstract execution environment parameter set.
+//!
+//! Parameter set is encoded as an opaque vector which structure depends on the execution
+//! environment itself (except for environment type/version which is always represented
+//! by the first element of the vector). Decoding to a usable semantics structure is
+//! done in `polkadot-node-core-pvf`.
+
+use crate::v2::{BlakeTwo256, HashT as _};
+use parity_scale_codec::{Decode, Encode};
+#[cfg(feature = "std")]
+use parity_util_mem::MallocSizeOf;
+use polkadot_core_primitives::Hash;
+use scale_info::TypeInfo;
+use sp_std::{ops::Deref, vec::Vec};
+
+/// # Execution environment parameter tags
+/// Values from 1 to 16 are reserved for system-wide parameters not related to any concrete
+/// execution environment.
+/// Environment type
+pub const EEPAR_01_ENVIRONMENT: u32 = 1;
+/// Logical stack limit, in stack items
+pub const EEPAR_17_STACK_LOGICAL_MAX: u32 = 17;
+/// Native stack limit, in bytes
+pub const EEPAR_18_STACK_NATIVE_MAX: u32 = 18;
+
+/// # Execution enviroment types
+/// Generic Wasmtime environment
+pub const EXEC_ENV_TYPE_WASMTIME_GENERIC: u32 = 0;
+
+/// Unit type wrapper around [`type@Hash`] that represents an execution parameter set hash.
+///
+/// This type is produced by [`ExecutorParams::hash`].
+#[derive(Clone, Copy, Encode, Decode, Hash, Eq, PartialEq, PartialOrd, Ord, TypeInfo)]
+pub struct ExecutorParamsHash(Hash);
+
+impl sp_std::fmt::Display for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		self.0.fmt(f)
+	}
+}
+
+impl sp_std::fmt::Debug for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		write!(f, "{:?}", self.0)
+	}
+}
+
+impl sp_std::fmt::LowerHex for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		sp_std::fmt::LowerHex::fmt(&self.0, f)
+	}
+}
+
+impl From<Hash> for ExecutorParamsHash {
+	fn from(hash: Hash) -> Self {
+		Self(hash)
+	}
+}
+
+impl From<[u8; 32]> for ExecutorParamsHash {
+	fn from(hash: [u8; 32]) -> Self {
+		Self(hash.into())
+	}
```

Comment:
This should be removed, otherwise the unit type wrapper is not that useful.

Reply:
Can you please elaborate? I used [`ValidationCodeHash`](https://github.com/paritytech/polkadot/blob/master/parachain/src/primitives.rs#L75) as a reference implementation.

Reply:
Hmm okay, I mixed this up with `CandidateHash`. However, I also think that `ValidationCodeHash` is wrong.

What I wanted to express, when you move stuff into unit type wrappers to have type level support, you should not make it too easy to construct the type. This means not having any kind of `From` implementation, because people otherwise may just call `into()` and then the entire reason for having this unit type wrapper is useless. I hope that explains it good enough, if not I can try to extend the explanation.

Reply:
Makes sense, and is not used in the current implementation anyway. I'll remove it.